trie-db/src/proof/eip1186.rs

```diff
@@ -0,0 +1,305 @@
+	proof: &'a[Vec<u8>],
+	) -> Result<(), VerifyError<'a, TrieHash<L>, CError<L>>>
+where
+L: TrieLayout,
+{
+	if key != nib && expected_value.is_none(){
+		return Ok(());
+	} else if key != nib {
+		return Err(VerifyError::NonExistingValue(key))
+	}
+	match_value::<L>(Some(data), key, expected_value, proof)
+}
+fn process_extension<'a,L>(
+	nib: &NibbleSlice,
+	handle: NodeHandle<'a>,
+	mut key: NibbleSlice<'a>,
+	expected_value: Option<&[u8]>,
+	proof: &'a[Vec<u8>],
+	) -> Result<(), VerifyError<'a,TrieHash<L>, CError<L>>>
+where
+L: TrieLayout,
+{
+	if !key.starts_with(nib) && expected_value.is_none() {
+		return Ok(());
+	} else if !key.starts_with(nib) {
+		return Err(VerifyError::NonExistingValue(key));
+	}
+	key.advance(nib.len());
+
+	match handle {
+		NodeHandle::Inline(encoded_node) => {
+			let root = L::Hash::hash(encoded_node);
+			process_node::<L>(&root, encoded_node, key, expected_value, proof)
+		},
+		NodeHandle::Hash(plain_hash) => {
+			let new_root = decode_hash::<L::Hash>(plain_hash).ok_or(VerifyError::HashDecodeError(plain_hash))?;
+			process_node::<L>(&new_root, &proof[0], key, expected_value, &proof[1..])
+		}
+	}
+}
+
+fn process_nibbledbranch<'a,L>(
+	nib: NibbleSlice,
+	children: [Option<NodeHandle<'a>>; 16],
+	maybe_data: Option<Value<'a>>,
+	mut key: NibbleSlice<'a>,
+	expected_value: Option<&[u8]>,
+	proof: &'a[Vec<u8>],
+	) -> Result<(), VerifyError<'a,TrieHash<L>, CError<L>>>
+where
+L: TrieLayout,
+{
+	if !key.starts_with(&nib) && expected_value.is_none() {
+		return Ok(());
+	} else if !key.starts_with(&nib) && expected_value.is_some(){
+		return Err(VerifyError::NonExistingValue(key))
+	}
+	key.advance(nib.len());
+
+	if key.is_empty() {
+		match_value::<L>(maybe_data, key, expected_value, proof)
+	} else {
+		match_children::<L>(children, key, expected_value, proof)
+	}
+}
+
+
+fn process_branch<'a,L>(
+	children: [Option<NodeHandle<'a>>; 16],
+	maybe_data: Option<Value<'a>>,
+	key: NibbleSlice<'a>,
+	expected_value: Option<&[u8]>,
+	proof: &'a[Vec<u8>],
+	) -> Result<(), VerifyError<'a, TrieHash<L>, CError<L>>>
+where
+L: TrieLayout,
+{
+	if key.is_empty() {
+		match_value::<L>(maybe_data, key, expected_value, proof)
+	} else {
+		match_children::<L>(children, key, expected_value, proof)
+	}
+}
+fn match_children<'a,L>(
+	children: [Option<NodeHandle<'a>>; 16],
+	mut key: NibbleSlice<'a>,
+	expected_value: Option<&[u8]>,
+	proof: &'a[Vec<u8>],
+	) -> Result<(), VerifyError<'a,TrieHash<L>, CError<L>>>
+where
+L: TrieLayout,
+{
+	match children.get(key.at(0) as usize) {
+		Some(Some(NodeHandle::Hash(hash))) => {
+			if proof.is_empty() {
+				Err(VerifyError::IncompleteProof)
+			} else {
+				key.advance(1);
+				let new_root = decode_hash::<L::Hash>(hash).ok_or(VerifyError::HashDecodeError(hash))?;
+				process_node::<L>(&new_root, &proof[0], key, expected_value, &proof[1..])
+			}
+		}
+		Some(Some(NodeHandle::Inline(encoded_node))) => {
+			key.advance(1);
+			let root = L::Hash::hash(encoded_node);
```

Comment:
This hash calculation is not needed.

Reply:
Sorry,
I am not sure I understand your comment. ``root`` is passed to ``process_node`` function the line after.
How can I avoid this hashing operation?

Reply:
IIReadC, in `process_node` this calculated hash is only compared again with encoded hash.
Making the input in `process_node` optional (or with an `enum { Inline, Node(H) }`) would work to avoid this check for inline node.